surge/blueprints.py

```diff
@@ -0,0 +1,94 @@
+import re
+from surge.projects import Project
+from surge.errors import SurgeMissingAttributeError
+
+
+class Blueprint(Project):
+    def __int__(self, **kwargs):
+        super().__init__(kwargs)
+
+    @classmethod
+    def _from_project(cls, project):
+        '''Converts a Project to a Blueprint. This is useful when we want to create a Project and then return
+        a Blueprint to the user.
+
+        Arguments:
+            project (Project): The project to convert to a Blueprint
+
+        Returns:
+            blueprint (Blueprint): A Blueprint derived from the given Project
+        '''
+        kwargs = project.__dict__
+        # A Project object will have already parsed created_at, so omit that and save current value to avoid errors.
+        created_at = kwargs.pop('created_at', None)
+        b = Blueprint(**project.__dict__)
+        b.created_at = created_at
+        return b
+
+    def required_data_fields(self):
+        '''
+        Returns all keys surrounded by {{}} from the fields_template string.
+
+        Returns:
+            matches (list): all the required data fields from fields_template
+            e.g. ['field1', 'field2']
+        '''
+        if not self.fields_template:
+            return []
+        pattern = r"{{(.*?)}}"
+        matches = re.findall(pattern, self.fields_template)
+        return matches
+
+    def create_new_batch(self, name):
+        '''
+        Create a new project from this blueprint. Once created, call create_tasks on the returned object
+        to assign tasks.
+
+        Arguments:
+            name (str): Name of the project
+
+        Returns:
+            blueprint (Blueprint): a created project from a blueprint project which can be assigned tasks and launched.
+        '''
+        if not name:
+            raise SurgeMissingAttributeError('name is required when creating a project from a template')
+
+        create_params = {'template_id': self.id, 'name': name}
+        project = Project.create(**create_params)
+        blueprint = Blueprint._from_project(project)
```

Comment:
Hmm - if you wanted to split the concept of a Blueprint project from a blueprint-derived project, could we split off the methods below this one, plus required_data_fields, into a separate class?

Reply:
It's a good question! I lingered over this for a while. But the real sticking point for me is the fact that list_blueprints was already returning Project objects. I was trying to paint "within the lines" as I saw them, and to my eye we already decided that Projects types were going to represent both Blueprints (or templates) and the Projects derived from them. If we decided that list_blueprints should return a non-Project Blueprint or Template object that would make sense to me, but it would be backward-incompatible which I assumed we didn't want in this PR.